src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -528,6 +540,66 @@ impl<'a, 'b, 'gcx, 'tcx> TypeVerifier<'a, 'b, 'gcx, 'tcx> {
         place_ty
     }
 
+    fn sanitize_promoted(&mut self, promoted_mir: &'b Mir<'tcx>, location: Location) {
+        // Determine the constraints from the promoted MIR by running the type
+        // checker on the promoted MIR, then transfer the constraints back to
+        // the main MIR, changing the locations to the provided location.
+
+        let main_mir = mem::replace(&mut self.mir, promoted_mir);
+        self.cx.mir = promoted_mir;
```

Comment:
hmm. It makes me a little sad that we've got this redundancy in the representation, in that I worry that other code might similarly swap in different MIR but omit one of the partnering references to that MIR (either because out of ignorance or because it doesn't have access to the outer `TypeVerifier`).

But I don't have a better answer for what to do here (well, apart from getting rid of the `self.mir` field in `TypeVerifier` and instead make those accesses go through `self.cx.mir`, potentially via a `mir()` method or via a `Deref` to the `self.cx`...)

Reply:
(Oh and of course I don't want to block *this* PR on such a change. Anything we can do to make this soundness fix more likely to be backported, I'm a fan of....)

Reply:
Well, this ended up happening anyway with the fix for the ICE.